### ISSUE TITLE
[Feature Request] Open In Finder on macOS

### ISSUE THREAD

**Issue by greggjaskiewicz** [FreeCAD/FreeCAD#12801]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

On macOS, usually every "document" driven application - and certainly FreeCAD should be classified to be in this category - has an option to either show the path the current file is in, by hovering or clicking on document icon - FreeCAD does not implement that level of integration with OS. 
Second option that is super useful - is to open the document in Finder (The file system browser). 
This helps, when trying to locate the directory, etc and once again - FreeCAD is missing this. 
<img width="401" alt="Screenshot 2024-03-07 at 13 35 50" src="https://github.com/FreeCAD/FreeCAD/assets/553583/6da1b910-4c19-4d7c-ba8c-7eea7280752b">

This context menu when right-clicking on the document - should contain a link to "Reveal in Finder" , or such. I'm sure QT has provisions for this built in, and this is a very much expected Pro feature of any document apps on macOS. 

<img width="427" alt="Screenshot 2024-03-07 at 13 38 06" src="https://github.com/FreeCAD/FreeCAD/assets/553583/4b21abe2-6da4-4c19-ad04-9d0def51459a">

### Full version info

```shell
OS: macOS 14.3.1
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.36135 (Git)
Build type: Release
Branch: main
Hash: e99220daf62069be653f73e720bcad9e0cd2fb97
Python 3.10.13, Qt 5.15.8, Coin 4.0.2, Vtk 9.2.6, OCC 7.6.3
Locale: C/Default (C)
Installed mods: 
  * AnimationFreeCAD 1.0.0
  * A2plus 0.4.62
  * sheetmetal 0.4.0
  * freecad_metal_workbench 0.0.1
  * Curves 0.6.23
```
```


### Subproject(s) affected?

None

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by iAmVaeVictus** [comment]
Agree. Every improvement of context menu options is a good thing, in my user (of MacOS) opinion. I think there are a lot of "context" right click choices that are absent that should be there, but we need to justify and specify each one to the developers to ensure they are implementing the things that people actually "need".  And, I don't currently understand FC well enough yet to know what's a reasonable ask for context sensitivity vs not in the complex world that is the Workbench-centric FreeCAD, with all it's inherent complexities.

**Comment by maxwxyz** [comment]
I think this would also benefit other OS.

**Comment by greggjaskiewicz** [comment]
100%

**Comment by luzpaz** [comment]
@greggjaskiewicz do you want to take a stab at this ?

**Comment by greggjaskiewicz** [comment]
Happy to work on it myself, in Qt there's a code :
void FileUtils::showInGraphicalShell(QWidget *parent, const QString &pathIn)
{
    const QFileInfo fileInfo(pathIn);
    // Mac, Windows support folder or file.
    if (HostOsInfo::isWindowsHost()) {
        const FileName explorer = Environment::systemEnvironment().searchInPath(QLatin1String("explorer.exe"));
        if (explorer.isEmpty()) {
            QMessageBox::warning(parent,
                                 QApplication::translate("Core::Internal",
                                                         "Launching Windows Explorer Failed"),
                                 QApplication::translate("Core::Internal",
                                                         "Could not find explorer.exe in path to launch Windows Explorer."));
            return;
        }
        QStringList param;
        if (!fileInfo.isDir())
            param += QLatin1String("/select,");
        param += QDir::toNativeSeparators(fileInfo.canonicalFilePath());
        QProcess::startDetached(explorer.toString(), param);
    } else if (HostOsInfo::isMacHost()) {
        QStringList scriptArgs;
        scriptArgs << QLatin1String("-e")
                   << QString::fromLatin1("tell application \"Finder\" to reveal POSIX file \"%1\"")
                                         .arg(fileInfo.canonicalFilePath());
        QProcess::execute(QLatin1String("/usr/bin/osascript"), scriptArgs);
        scriptArgs.clear();
        scriptArgs << QLatin1String("-e")
                   << QLatin1String("tell application \"Finder\" to activate");
        QProcess::execute(QLatin1String("/usr/bin/osascript"), scriptArgs);
    } else {
        // we cannot select a file here, because no file browser really supports it...
        const QString folder = fileInfo.isDir() ? fileInfo.absoluteFilePath() : fileInfo.filePath();
        const QString app = UnixUtils::fileBrowser(ICore::settings());
        QProcess browserProc;
        const QString browserArgs = UnixUtils::substituteFileBrowserParameters(app, folder);
        bool success = browserProc.startDetached(browserArgs);
        const QString error = QString::fromLocal8Bit(browserProc.readAllStandardError());
        success = success && error.isEmpty();
        if (!success)
            showGraphicalShellError(parent, app, error);
    }
}

Refactoring and shoving it somewhere in the Tree.cpp actions, would probably work. I'm happy to look at it myself, but might take me a bit to get there :) Never done FreeCAD dev before and I can only compile it on my linux laptop - macOS compilation doesn't seem to be made easier by cmake. .. Basically looks like little love was given to macOS support, which is a shame.

**Comment by luzpaz** [comment]
Thanks!

> Basically looks like little love was given to macOS support, which is a shame.  

Just like many other libregraphic foss applications, macOS development lags behind. Probably due to the OS's lack of hospitality toward FLOSS in general.

**Comment by greggjaskiewicz** [comment]
@luzpaz exploring how to do it currently. Wish more people were on discord to talk. :)

**Comment by greggjaskiewicz** [comment]
I'm having this problem https://forum.freecad.org/viewtopic.php?p=749556#p749556 when building, which blocks me from staring working on this @luzpaz @maxwxyz

**Comment by luzpaz** [comment]
Forum thread shows it's not a FC related issue. Did you end up building successfully?

**Comment by greggjaskiewicz** [comment]
yes, I got it building fine now - required me to build my own yaml-cpp from scratch etc.

**Comment by luzpaz** [comment]
Is the yaml issue related to #13264 ?

**Comment by greggjaskiewicz** [comment]
quite possibly, although that issue was for linux

**Comment by luzpaz** [comment]
Darn, forgot about this one.

**Comment by luzpaz** [comment]
@hyarion any interest in this ?

**Comment by hyarion** [comment]
@luzpaz maybe! :D
It is a good task to save for a new dev though. I think we can drop macOS tag on it as it would be useful on other systems as well.

I would recommend calling it something like "Reveal in Finder" on mac and "Open File Location" for other systems.
